Change the PAKE output function to return a key

athoelke opened the pull request:
* Replace `psa_pake_get_implicit_key()` with `psa_pake_get_shared_key()`
* Add key confirmation attribute to the PAKE cipher suite

Fixes #100
Fixes #86

athoelke commented:
Some feedback would be very welcome from @silabs-Kusumit, @silabs-hannes, @oberon-sk, and @yanesca; given your involvement in the PAKE API development.

oberon-sk commented:
> * Replace `psa_pake_get_implicit_key()` with `psa_pake_get_shared_key()`
> * Add key confirmation attribute to the PAKE cipher suite

lgtm

> 1. Should the entire secret be contained in a single key output, or can it be split across multiple output keys?

As far as we understand, the goal of this PR is to export a shared secret and defer any key derivation of one or multiple keys into subsequent calls independent from the PAKE APIs. Then, the entire secret needs to be exported.

> 5. Which position should the key attributes parameter take in `psa_pake_get_shared_key()`? - there is no ideal answer as the precedents in the API are not consistent.

The `operation` should always be the first parameter, we regard `psa_key_derivation_output_key()` as an outlier.

athoelke commented:
> > * Replace `psa_pake_get_implicit_key()` with `psa_pake_get_shared_key()`
> > * Add key confirmation attribute to the PAKE cipher suite
> 
> lgtm

Thank you for your review

> > 1. Should the entire secret be contained in a single key output, or can it be split across multiple output keys?
> 
> As far as we understand, the goal of this PR is to export a shared secret and defer any key derivation of one or multiple keys into subsequent calls independent from the PAKE APIs. Then, the entire secret needs to be exported.

The concern is whether there are protocols, or PAKEs that expect the output to be split into two or more pieces as keys. For example, the way that SPAKE2+-draft02 splits the output of `Hash(TT)` into `Ka` and `Ke`. Note that is not a use case we need to support, and Matter's use of SPAKE2+ only uses the entire confirmed PAKE output as a key derivation key.

> > 5. Which position should the key attributes parameter take in `psa_pake_get_shared_key()`? - there is no ideal answer as the precedents in the API are not consistent.
> 
> The `operation` should always be the first parameter, we regard `psa_key_derivation_output_key()` as an outlier.

We have come to the same conclusion.

oberon-sk commented:
> The concern is whether there are protocols, or PAKEs that expect the output to be split into two or more pieces as keys. For example, the way that SPAKE2+-draft02 splits the output of Hash(TT) into Ka and Ke. Note that is not a use case we need to support, and Matter's use of SPAKE2+ only uses the entire confirmed PAKE output as a key derivation key.

We are not aware of a use case where this would be required either.

athoelke commented:
> 3. What key types should we permit or deny? - it seems that this is not appropriate for extracting keys that can require a variable amount of keying material (e.g. any DES, FFDH key pairs, Weierstrass ECC key pairs, RSA key pairs)

We would like to make this 'derivation keys only' (similar to the `psa_key_agreement()` PR #101), but need confidence that there are no use cases where the output is used directly for encryption, key wrapping (might need block-cipher-key), or AES-CMAC-based key derivation (would need an AES key).

> 4. Should the default key confirmation be: 'unconfirmed', 'confirmed', or 'whatever the default is for the selected algorithm'?

Currently leaning towards a preference of 'confirmed' by default (following a 'secure by default' principle), which means that algorithms like J-PAKE, without a confirmation phase, would require the application to explicitly request an unconfirmed key.

athoelke commented:
Added a missing update to the J-PAKE graphic, and reversed the definition of the key confirmation constants, so `PSA_PAKE_CONFIRMED_KEY` is now the default.

silabs-Kusumit commented on doc/ext-pake/api/pake.rst:
```diff
@@ -343,9 +346,13 @@ A PAKE primitive is required when constructing a PAKE cipher-suite object, `psa_
-A PAKE algorithm uses a specific cryptographic primitive for key establishment, specified using a `PAKE primitive <pake-primitive>`. PAKE algorithms also require a cryptographic hash algorithm, which is agreed between the participants.
+Most PAKE algorithms have parameters that must be specified by the application. These parameters include:
+
+*   The cryptographic primitive used for for key establishment, specified using a `PAKE primitive <pake-primitive>`.
```
Typo

athoelke replied:
Sorry, I can't seem to see the problem. What am I missing?

silabs-Kusumit replied:
> The cryptographic primitive used for for key establishment

Small thing

athoelke replied:
Thank you! The hardest to see error in things you have written yourself....

silabs-Kusumit commented:
> 1. Should the entire secret be contained in a single key output, or can it be split across multiple output keys?

If the key type of the output will be derivation only then multiple output keys are not needed.

> Added a missing update to the J-PAKE graphic, and reversed the definition of the key confirmation constants, so `PSA_PAKE_CONFIRMED_KEY` is now the default.

lgtm

> 2\. Related to (1): Do we need to provide something that evaluates the size of the shared key output?

If we are returning only a single output key then this may not be needed.

athoelke commented:
> 3. What key types should we permit or deny? - it seems that this is not appropriate for extracting keys that can require a variable amount of keying material (e.g. any DES, FFDH key pairs, Weierstrass ECC key pairs, RSA key pairs)

Even if we expect that the output should be for derivation only: we should permit implementations to output AES keys (as long as they use all of the shared secret?), as we expect some implementations will provide AES-CMAC based KDFs (see #106).

yanesca commented on doc/ext-pake/api.db/psa/crypto-pake.h:
```diff
@@ -27,22 +28,29 @@ typedef uint8_t psa_pake_step_t;
-                                       psa_key_derivation_operation_t *output);
+psa_status_t psa_pake_get_shared_key(psa_pake_operation_t *operation,
+                                     const psa_key_attributes_t * attributes,
+                                     psa_key_id_t * key);
```
The situation here is very similar to the case of [key agreement](https://armmbed.github.io/mbed-crypto/html/api/ops/ka.html) schemes. There we have two APIs: `psa_raw_key_agreement()` and `psa_key_derivation_key_agreement()`. Outputing a key derivation object achieves the same as the latter. Outputing a key instead is somewhere between the two: the Crypto API implementation doesn't have full control over the use of the output, but doesn't release the output to the user either.

Outputting a key derivation object has the advantage that the implementation knows about the PAKE when inputting the key to the derivation object and can prevent misuse better or even adjust serialisation format if needed. If the PAKE output is released as a key, the only information the implementation can rely on is the length.

Why do we need to switch to outputting a key?

athoelke replied:
> Why do we need to switch to outputting a key?

See #86. Use cases that use separate derivations to create multiple keys from the output of the PAKE. The existing API cannot be used in such systems, and maintaining the existing API alongside a new one to return a key does not appear to deliver substantial benefit for the application or implementation.

yanesca commented:
> Even if we expect that the output should be for derivation only: we should permit implementations to output AES keys (as long as they use all of the shared secret?), as we expect some implementations will provide AES-CMAC based KDFs (see https://github.com/ARM-software/psa-api/issues/106).

I think we should permit implementations to output any type of key that we might need for key derivation. If I understand the proposal in #106 correctly, this would mean `PSA_KEY_TYPE_DERIVE`, `PSA_KEY_TYPE_HMAC`, `PSA_KEY_TYPE_AES` and for the last two the key policy would be set to `PSA_ALG_SP800_108_COUNTER_HMAC ` and `PSA_ALG_SP800_108_COUNTER_CMAC` respectively.

yanesca commented:
> Should the entire secret be contained in a single key output, or can it be split across multiple output keys?

I think most of the PAKEs arrive at one shared secret and the output should be a single key. If a higher level protocol splits that before deriving keys, the partitioning of the key material should be part of the key derivation algorithm. For PAKEs that output several keys (like OPAQUE) we should either add a new API or require the application to call the key output function once more (and define which call produces which key), but we shouldn't output a concatenated key.

> Related to (1): Do we need to provide something that evaluates the size of the shared key output?

If we put the output in a key, the caller doesn't need to know the size upfront and can get it from the key attributes later, so I don't think we need to provide anything extra here.

athoelke commented:
> I think most of the PAKEs arrive at one shared secret and the output should be a single key. If a higher level protocol splits that before deriving keys, the partitioning of the key material should be part of the key derivation algorithm.

That is also my opinion, all other things considered.

> For PAKEs that output several keys (like OPAQUE) we should either add a new API or require the application to call the key output function once more (and define which call produces which key), but we shouldn't output a concatenated key.

If I understand you correctly, you suggest that for algorithms like OPAQUE, we might decide that the application must call `psa_pake_get_shared_key()` twice. However, the application does _not_ control how the output is split between those two calls: that is a product of the algorithm itself; in contrast to the idea in (1) where the application could control the splitting of the PAKE output into multiple key objects.

athoelke commented:
Rebased (following merge of #115), and updated the specification of key types that `psa_pake_get_shared_key()` can output (see https://github.com/ARM-software/psa-api/pull/114/commits/9d4f2c06cc737e029d19d3fb24e777daef8ea761).

Some PAKEs (e.g. J-PAKE) produce biased output, that is not suitable for use as an encryption key. Therefore:

* Only require that `PSA_KEY_TYPE_DERIVE` and `PSA_KEY_TYPE_HMAC` (see #123) are valid for all PAKE algorithms
* Permit additional key types (block-cipher keys) for PAKE algorithms that produce a pseudo-random shared secret.

yanesca commented on doc/ext-pake/api/pake.rst:
```diff
@@ -1084,8 +1088,14 @@ Multi-part PAKE operations
+        *   The key type is required.
+            All PAKE algorithms can output a key of type :code:`PSA_KEY_TYPE_DERIVE` or :code:`PSA_KEY_TYPE_HMAC`.
+            PAKE algorithms that produce a pseudo-random shared secret, can also output block-cipher key types, for example :code:`PSA_KEY_TYPE_AES`.
+
```
Is this empty line intentional?

athoelke replied:
In this case, I think not, it was probably an artefact of cut and pasting the text below the blank line.

yanesca reviewed (approved):
LGTM.

silabs-Kusumit commented:
LGTM

athoelke commented:
Rebased and merged.
